dlsonic/dlsonic.C

```diff
@@ -0,0 +1,696 @@
+{
+    std::optional<std::pair<di::Instruction, uint32_t>> inst;
+    auto inst_rdi =  locateAssignmentInstruction( Dyninst::x86_64::irdi, b, obj, fn );
+    auto inst_edi = locateAssignmentInstruction( Dyninst::x86_64::iedi, b, obj, fn );
+    
+    if ( inst_rdi.has_value() ) {
+        inst = inst_rdi;
+    } else if ( inst_edi.has_value() ) {
+        inst = inst_edi;
+    } else {
+        // this means we couldn't find any assignment to RDI
+        // typically this will imply that a dlsym with nullptr
+        // often this appears as xor rdi rdi.
+        return;
+    }
+
+    auto val = inst.value();
+    auto allNodes = doSlice( obj, val.first, val.second, fn, b, Dyninst::x86_64::irdi );
+
+    auto index = GlobalData::Instance().calldetails.back().id;
+
+    // now we look at constant assignment case
+    // instead of the handle for dlsym, RTLD_NEXT / RTLD_DEFAULT are often passed
+    // we need to keep track of these 
+    if ( allNodes.size() == 1 ) {
+        auto sliceNode = dynamic_cast<Dyninst::SliceNode*>( allNodes.back().get() );
+        auto tgt = sliceNode->assign()->insn();
+        if ( tgt.getOperation().getID() == e_mov ) {
+            auto result = tgt.getOperand(1).getValue()->eval();
+            if ( result.defined ) {
+                // Internally dyninst is storing void* values as unsigned 4 byte values
+                auto val = result.convert<int32_t>();
+                if ( val == reinterpret_cast<int64_t>( RTLD_DEFAULT ) ) {
+                    GlobalData::Instance().calldetails[index-1].htype = GlobalData::DlHandleType::CONST_RTLDDEFAULT;
+                    Stats::Instance().dlsymWithRTLD_DEFAULT++;
+                } else if (val == reinterpret_cast<int64_t>( RTLD_NEXT ) ) {
+                    GlobalData::Instance().calldetails[index-1].htype = GlobalData::DlHandleType::CONST_RTLDNEXT; 
+                    Stats::Instance().dlsymWithRTLD_NEXT++;
+                } else {
+                    std::cerr << "Found an unknown const handle found for dlsym call: " << val << std::endl;
+                    GlobalData::Instance().calldetails[index-1].htype = GlobalData::DlHandleType::CONST_UNKNOWN; 
+                }
+                Stats::Instance().dlsymWithConstHandle++;
+            }
+        }
+    }
+
+    GlobalData::Instance().dlsymIndex2RDISlice[index] = allNodes; 
+}
+
+bool sliceNodeContainsRAXIn( Dyninst::Node::Ptr node )
+{
+    auto insn = dynamic_cast<Dyninst::SliceNode*>(node.get())->assign()->insn();
+
+    std::vector<di::Operand> instOpr;
+    insn.getOperands( instOpr );
+
+    for ( auto op: instOpr ) {
+        std::set<di::RegisterAST::Ptr> regSet;
+        op.getReadSet( regSet );
+        for ( auto w: regSet ) {
+            if ( w->getID() == Dyninst::x86_64::irax ) {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+
+} // end anonymous namespace
+
+int main( int argc, char* argv[] )
+{
+    std::string execName = argv[1];
```

Comment:
If there are no arguments this undefined as constructing a string from a nullptr is undefined.  Before this test is argc != 2 and if so print usage message and exit 1.